plane/src/proxy/rewriter.rs

```diff
@@ -61,6 +77,32 @@ impl RequestRewriter {
         &self.bearer_token
     }
 
+    pub fn get_host_header(&self) -> Option<&HeaderValue> {
+        self.parts.headers.get(HOST)
+    }
+
+    pub fn get_subdomain(
+        &self,
+        cluster: &ClusterName,
+    ) -> Result<Option<&str>, RequestRewriterError> {
+        let subdomain = self
+            .get_host_header()
+            .map(|h| h.to_str())
+            .transpose()? // failure to parse (non-ascii)
+            .map(|h| {
+                h.strip_suffix(cluster.as_str())
+                    .ok_or(RequestRewriterError::InvalidHostHeader)
+            })
+            .transpose()? // doesn't end in cluster name
+            .filter(|s| !s.is_empty()) // filter out here if there's no subdomain
+            .map(|s| {
+                s.strip_suffix('.')
+                    .ok_or(RequestRewriterError::InvalidHostHeader)
+            })
```

Comment:
I think the order here should be flipped, we want to check if the subdomain is empty after removing the `.`. I guess it doesn't really matter since the client would have to do something intentionally weird to send a host header of the form `.{cluster}`, and we'll still reject the request, but assuming that it's the ~same code complexity to handle that case by rejecting the domain I think it's worth doing that.